scripts/sign_encrypt.py

```diff
@@ -29,391 +61,845 @@ def int_parse(str):
+
+        self.chosen_hash = hashes.SHA256()
+        self.digest_len = self.chosen_hash.digest_size
+
+    def __pack_img(self, img_type, sign_algo):
+        import struct
+
+        self.sig_algo = sign_algo
+        self.img_type = img_type
+        self.shdr = struct.pack('<IIIIHH', SHDR_MAGIC, img_type, len(self.img),
+                                sig_tee_alg[sign_algo], self.digest_len,
+                                self.sig_len)
+
+    def __calc_digest(self):
+        from cryptography.hazmat.backends import default_backend
+        from cryptography.hazmat.primitives import hashes
+
+        h = hashes.Hash(self.chosen_hash, default_backend())
+        h.update(self.shdr)
+        if hasattr(self, 'ta_uuid'):
+            h.update(self.ta_uuid)
+            h.update(self.ta_version)
+        if hasattr(self, 'ehdr'):
+            h.update(self.ehdr)
+            h.update(self.nonce)
+            h.update(self.tag)
+        h.update(self.img)
+        return h.finalize()
+
+    def encrypt_ta(self, enc_key, key_type, sig_algo, uuid, ta_version):
+        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+        import struct
+
+        self.img = self.inf
+
+        cipher = AESGCM(bytes.fromhex(enc_key))
+        self.nonce = os.urandom(NONCE_SIZE)
+        out = cipher.encrypt(nonce, self.img, None)
+        self.ciphertext = out[:-TAG_SIZE]
+        # Authentication Tag is always the last bytes
+        self.tag = out[-TAG_SIZE:]
+
+        enc_algo = enc_tee_alg['TEE_ALG_AES_GCM']
+        flags = enc_key_type[key_type]
+        self.ehdr = struct.pack('<IIHH', enc_algo, flags, len(self.nonce),
+                                len(self.tag))
+
+        self.__pack_img(SHDR_ENCRYPTED_TA, sig_algo)
+        self.ta_uuid = uuid.bytes
+        self.ta_version = struct.pack('<I', ta_version)
+        self.img_digest = self.__calc_digest()
+
+    def set_bootstrap_ta(self, sig_algo, uuid, ta_version):
+        import struct
+
+        self.img = self.inf
+        self.__pack_img(SHDR_BOOTSTRAP_TA, sig_algo)
+        self.ta_uuid = uuid.bytes
+        self.ta_version = struct.pack('<I', ta_version)
+        self.img_digest = self.__calc_digest()
+
+    def set_subkey(self, sign_algo, name, uuid, subkey_version, max_depth,
+                   name_size):
+        from cryptography.hazmat.primitives.asymmetric import rsa
+        import struct
+
+        self.subkey_name = name
+
+        subkey_key = load_asymmetric_key_img(self.inf)
+        if isinstance(subkey_key, rsa.RSAPrivateKey):
+            subkey_pkey = subkey_key.public_key()
+        else:
+            subkey_pkey = subkey_key
 
-    img_size = len(img)
+        if max_depth is None:
+            if hasattr(self, 'previous_max_depth'):
+                max_depth = self.previous_max_depth - 1
+            else:
+                max_depth = 0
+        if max_depth < 0:
+            raise Exception("Cannot use max_depth {}".format(max_depth))
```

Comment:
General comment: since we're using Python 3, f-strings could help make things more compact (`f"Cannot use max_depth {max_depth}"`).

Reply:
OK if we take this in a new commit on top?

Reply:
Whatever is easier for you. I don't mind reviewing the whole series again if you prefer.

Reply:
I read that f-strings are a python 3.6 thing and we actually don't use f-strings in any python-script in this git, yet. I'd rather have that concept introduced in a separate PR i you don't mind.

Reply:
> we actually don't use f-strings in any python-script in this git, yet

Right, in fact I think I am the one who introduced them in the (now deleted) file `dump_ta_header.py`. Let's deal with that later especially since it requires Python 3.6+.